crates/ironrdp-rdpdr/src/pdu/efs.rs

```diff
@@ -127,67 +25,81 @@ pub enum VersionAndIdPduKind {
     pub version_major: u16,
     pub version_minor: u16,
     pub client_id: u32,
     pub kind: VersionAndIdPduKind,
 }
 
 impl VersionAndIdPdu {
-    /// The size of the PDU without the header
-    const HEADERLESS_SIZE: usize = (size_of::<u16>() * 2) + size_of::<u32>();
-
-    fn header(&self) -> SharedHeader {
-        match self.kind {
-            VersionAndIdPduKind::ClientAnnounceReply => SharedHeader {
-                component: Component::RdpdrCtypCore,
-                packet_id: PacketId::CoreClientidConfirm,
-            },
-            VersionAndIdPduKind::ServerAnnounceRequest => SharedHeader {
-                component: Component::RdpdrCtypCore,
-                packet_id: PacketId::CoreServerAnnounce,
-            },
+    const FIXED_PART_SIZE: usize = (size_of::<u16>() * 2) + size_of::<u32>();
+
+    pub fn new_client_announce_reply(req: VersionAndIdPdu) -> PduResult<Self> {
+        if req.kind != VersionAndIdPduKind::ServerAnnounceRequest {
+            return Err(invalid_message_err!(
+                "VersionAndIdPdu::new_client_announce_reply",
+                "VersionAndIdPduKind",
+                "invalid value"
+            ));
         }
-    }
 
-    pub fn decode(src: &mut ReadCursor, kind: VersionAndIdPduKind) -> PduResult<Self> {
-        ensure_size!(ctx: kind.name(), in: src, size: Self::HEADERLESS_SIZE);
         Ok(Self {
-            version_major: src.read_u16(),
-            version_minor: src.read_u16(),
-            client_id: src.read_u32(),
-            kind,
+            version_major: VERSION_MAJOR,
+            version_minor: VERSION_MINOR_12,
+            client_id: req.client_id,
+            kind: VersionAndIdPduKind::ClientAnnounceReply,
         })
     }
-}
 
-impl PduEncode for VersionAndIdPdu {
-    fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
-        ensure_size!(ctx: self.name(), in: dst, size: self.size());
-        self.header().encode(dst)?;
+    pub fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        ensure_size!(ctx: self.name(), in: dst, size: Self::FIXED_PART_SIZE);
         dst.write_u16(self.version_major);
         dst.write_u16(self.version_minor);
         dst.write_u32(self.client_id);
         Ok(())
     }
 
-    fn name(&self) -> &'static str {
+    pub fn decode(header: SharedHeader, src: &mut ReadCursor) -> PduResult<Self> {
+        let kind = match header.packet_id {
+            PacketId::CoreServerAnnounce => VersionAndIdPduKind::ServerAnnounceRequest,
+            PacketId::CoreClientidConfirm => VersionAndIdPduKind::ClientAnnounceReply,
+            _ => {
+                return Err(invalid_message_err!(
+                    "VersionAndIdPdu::decode",
+                    "PacketId",
+                    "invalid value"
+                ))
+            }
+        };
+
+        ensure_size!(ctx: kind.name(), in: src, size: Self::FIXED_PART_SIZE);
+
+        Ok(Self {
+            version_major: src.read_u16(),
```

Comment:
[nit] This is definitely a nitpicking by me, you could ignore this comment if you think otherwise, but In my opinion struct field initialization should not contain code, which depends on execution order -- I always use additional `let`s for such cases:

```rust
let version_major = src.read_u16();
let version_minor = src.read_u16();
let client_id = src.read_u32();

Ok(Self {
    version_major,
    version_minor,
    client_id,
    kind,
})
```

But in case of PDUs I guess it is not so important, as it is hard to mix field ordering during consecutive refactoring due to immutable nature of PDUs